gradle-jdks-setup/src/main/java/org/gradle/wrapper/GradleWrapperMain.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.wrapper;
+
+import com.palantir.gradle.jdks.common.CommandRunner;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.UnaryOperator;
+
+@SuppressWarnings("BanSystemOut")
+public final class GradleWrapperMain {
+
+    private GradleWrapperMain() {}
+
+    public static void main(String[] args) throws Exception {
+        configureJdkAutomanagement();
+
+        // Delegate back to the original GradleWrapperMain implementation
+        ClassLoader loader = new OriginalGradleWrapperLoader();
+        Class<?> origGradleWrapperMainClass = loader.loadClass("org.gradle.wrapper.OrigGradleWrapper");
+        Method method = origGradleWrapperMainClass.getMethod("main", String[].class);
+        method.invoke(null, new Object[] {args});
+    }
+
+    private static void configureJdkAutomanagement() throws IOException {
+        // Delegate to gradle-jdks-setup.sh to install the JDK
+        // TODO(crogoz): maybe use java to do the setup
+        Path projectHome = projectHome();
+        CommandRunner.run(List.of("sh", "./gradle/gradle-jdks-setup.sh"), Optional.of(projectHome.toFile()));
```

Comment:
We should probably add a minor optimisation at some point to not do this if Gradle has been invoked from `./gradlew` - setting an environment variable or something we can check.